File: app/src/main/java/com/salino/sali - Copy/data/remote/source/ShoppingRemoteDataSource.kt
package com.salino.sali.data.remote.source

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import com.salino.sali.data.model.ShoppingItem
import kotlinx.coroutines.tasks.await
import javax.inject.Inject

class ShoppingRemoteDataSource @Inject constructor(
    private val firestore: FirebaseFirestore
) {
    private fun itemsCollection(householdId: String) = firestore.collection("households")
        .document(householdId)
        .collection("items")

    fun listenToItems(householdId: String, onItems: (List<ShoppingItem>) -> Unit): ListenerRegistration {
        return itemsCollection(householdId)
            .orderBy("updatedAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, _ ->
                val items = snapshot?.documents?.mapNotNull { doc ->
                    doc.toObject(ShoppingItem::class.java)?.copy(id = doc.id)
                } ?: emptyList()
                onItems(items)
            }
    }

    suspend fun upsertItem(householdId: String, item: ShoppingItem) {
        itemsCollection(householdId).document(item.id).set(item).await()
    }

    suspend fun deleteItem(householdId: String, itemId: String) {
        itemsCollection(householdId).document(itemId).delete().await()
    }

    suspend fun getItem(householdId: String, itemId: String): ShoppingItem {
        val doc = itemsCollection(householdId).document(itemId).get().await()
        return doc.toObject(ShoppingItem::class.java)?.copy(id = doc.id)
            ?: throw IllegalStateException("Item not found")
    }
}

File: ios/Remote/ShoppingRemoteDataSource.swift
import Foundation
import FirebaseFirestore

final class ShoppingRemoteDataSource {
    private let firestore: Firestore?

    init(firestore: Firestore? = configuredFirestore()) {
        self.firestore = firestore
    }

    private func itemsCollection(_ householdId: String) throws -> CollectionReference {
        guard let firestore else { throw FirebaseConfigurationError.missingGoogleServiceInfo }
        return firestore.collection(FirestoreCollections.households)
            .document(householdId)
            .collection(FirestoreCollections.items)
    }

    func listenToItems(householdId: String, onItems: @escaping ([ShoppingItem]) -> Void) -> ListenerRegistration {
        guard let collection = try? itemsCollection(householdId) else {
            onItems([])
            return NoopListenerRegistration()
        }
        return collection
            .order(by: "updatedAt", descending: true)
            .addSnapshotListener { snapshot, _ in
                let items = snapshot?.documents.map { ShoppingItem(documentId: $0.documentID, data: $0.data()) } ?? []
                onItems(items)
            }
    }

    func upsertItem(householdId: String, item: ShoppingItem) async throws {
        try await itemsCollection(householdId).document(item.id).setDataAsync(item.firestoreData)
    }

    func deleteItem(householdId: String, itemId: String) async throws {
        try await itemsCollection(householdId).document(itemId).deleteAsync()
    }

    func getItem(householdId: String, itemId: String) async throws -> ShoppingItem {
        let snapshot = try await itemsCollection(householdId).document(itemId).getDocumentAsync()
        guard snapshot.exists else { throw FirebaseConfigurationError.documentNotFound }
        return ShoppingItem(documentId: snapshot.documentID, data: snapshot.dataOrEmpty)
    }
}
